File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/require-doc-content.ts
#!/usr/bin/env bun
import type { Finding, HarnessCheckRule, HarnessManifest, RuleContext } from "../harness-check-rule";

/**
 * Require specified content in documentation files.
 */
export class RequireDocContentRule implements HarnessCheckRule {
  static readonly category = "requireDocContent";

  constructor(private readonly ctx: RuleContext) {}

  applies(manifest: HarnessManifest): boolean {
    const section = manifest.requireDocContent;
    if (typeof section !== "object" || section === null) {
      return false;
    }
    const enabled = (section as { enabled?: unknown }).enabled;
    if (enabled === false) {
      return false;
    }
    const parameters = this.ctx.readJsonObject((section as Record<string, unknown>).parameters);
    const checks = parameters.checks;
    return Array.isArray(checks) && checks.length > 0;
  }

  validate(_root: string, manifest: HarnessManifest): readonly Finding[] {
    const section = this.ctx.readJsonObject(manifest.requireDocContent);
    const parameters = this.ctx.readJsonObject(section.parameters);
    const checks = parameters.checks;
    if (!Array.isArray(checks)) {
      return [];
    }
    return checks.flatMap((check) => {
      if (typeof check !== "object" || check === null) {
        return [];
      }
      const checkObj = check as Record<string, unknown>;
      const files = this.ctx.readStringArray(checkObj.files);
      const containsAll = this.ctx.readStringArray(checkObj.containsAll);
      const failureMessage = typeof checkObj.failureMessage === "string" ? checkObj.failureMessage : "";

      const combinedText = files.map((f) => this.ctx.read(f)).join("\n");
      const hasAllSubstrings = containsAll.every((substring) => combinedText.includes(substring));

      return !hasAllSubstrings && failureMessage
        ? [
            {
              severity: this.ctx.severityOf(manifest, RequireDocContentRule.category),
              category: RequireDocContentRule.category,
              message: failureMessage,
            },
          ]
        : [];
    });
  }
}

File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/require_doc_content.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""Require doc content rule."""

from pathlib import Path

from harness_check_rule import Finding, HarnessCheckRule

from ._utils import is_safe_file, read_text, severity_for


class RequireDocContentRule(HarnessCheckRule):
    """Validate requireDocContent check."""

    category = "requireDocContent"

    def applies(self, manifest: dict) -> bool:
        """Check if this rule applies to the manifest."""
        section = manifest.get(self.category)
        if not isinstance(section, dict):
            return False
        enabled = section.get("enabled", True)
        return enabled is not False

    def validate(self, root: Path, manifest: dict) -> list[Finding]:
        """Validate requireDocContent check."""
        section = manifest.get(self.category, {})
        if not isinstance(section, dict):
            return []
        params = section.get("parameters", {})
        if not isinstance(params, dict):
            return []
        checks = params.get("checks", [])
        if not isinstance(checks, list):
            return []
        result = []
        for check in checks:
            if not isinstance(check, dict):
                continue
            if not isinstance(check.get("files"), list):
                continue
            if not isinstance(check.get("containsAll"), list):
                continue
            if not isinstance(check.get("failureMessage"), str):
                continue
            files = check.get("files", [])
            contains_all = check.get("containsAll", [])
            failure_message = check.get("failureMessage")
            combined_text = "\n".join(
                read_text(root / f)
                for f in files
                if isinstance(f, str) and is_safe_file(root / f)
            )
            if not all(
                isinstance(s, str) and s in combined_text
                for s in contains_all
            ):
                result.append(Finding(
                    severity_for(manifest, self.category),
                    self.category,
                    failure_message,
                ))
        return result
